src/helpers/qownnotesmarkdownhighlighter.cpp

```diff
@@ -127,23 +118,36 @@ void QOwnNotesMarkdownHighlighter::highlightMarkdown(const QString& text) {
         highlightAdditionalRules(_highlightingRulesPre, text);
 
         // needs to be called after the horizontal ruler highlighting
-        highlightHeadline(text);
+        const QString &next = currentBlock().next().text();
+        if (text.contains(QStringLiteral("# ")) || text.contains(QStringLiteral("## ")) ||
```

Comment:
I guess you `startsWith` would be more correct instead of all `contains`.
And instead of all the `# ` checks a regular expression that checks one to six occurrences would also do, wouldn't it?

Reply:
I was having problems with startsWith initially. Just tired it again, and it seems to be working fine. No differences in performance.

Using regex is a lot more expensive. It would be more concise and clean I agree, but the performance penalty is going to be much bigger.

As of now, a 16K line note (which is almost equal to a 400 page book) is loading in almost 0.4s with spellchecking enabled(without autodetect).  With autodetect it's 1 sec.

This is a lot faster than many other applications, and most if not all of the Qt based markdown readers and certainly faster than all the electron based apps. (Have tested ghostwriter, focus writer, marktext, Libreoffice, typora). Not to mention that we are doing full markdown highlighting, and not skipping anything.

The slowest function of them all is `highlightBrokenNotesLink()` followed by `highlightHeadline()` and then `highlight-additionalRulesAfter`.

We can speed this up even more, but for that we will have to seperate functionality into smaller functions. For e.g., taking some options out of `highlightAdditionalRulesAfter` and putting them in seperate functions and then placing checks on them to decide whether it needs to be run or not. Separating functionality makes it easier to place checks for individual elements (tables, lists etc.) Same could be done with `headings` and `brokenNotelinks`.

Reply:
You are doing an awesome job here! I'm always amazed that improvements are made!

Reply:
;)

Reply:
I meant "what improvements are made" :smile:

Reply:
Both sentences are correct I think :rofl: